### PR TITLE
build(ci): Add basic linting on Travis for Python 3 syntax errors

### PR DESCRIPTION
**Pull request by cclauss**
Also finds undefined names, etc.  @untitaker To help in the Python 3 porting efforts.

These tests are run in __allow_failures__ mode for now.

### CODE REVIEW

**Comment by untitaker** [comment]
@cclauss I don't believe we want to add any new linters right now, particularly not in a different configuration than on Python 2

**Comment by wedamija** [comment]
This is a decent idea since we're working on python 3 compatibility this quarter. I'm not going to approve/merge it just yet, but once the project is moving along we might do so.

**Comment by joshuarli** [comment]
@cclauss It'd probably be a good idea to pin to the same flake8 version we use for 2.7. Also, we'd need https://github.com/getsentry/sentry-flake8 as it runs some helpful custom checks. Installing `sentry-flake8==0.1.1` rather than `flake8` can hit two birds with one stone here.

**Comment by cclauss** [comment]
getsentry/sentry-flake8#5

**Comment by cclauss** [comment]
#14923 is the only issue that remains.

**Comment by cclauss** [comment]
> I don't believe we want to add any new linters right now

Why not?!?  With < 75 days left until legacy Python EOL, why not put guards in place to stop contributors from breaking Python compatibility?  I have been [fixing these errors in this codebase](https://github.com/getsentry/sentry/pulls?page=2&q=is%3Apr+author%3Acclauss+is%3Aclosed) for two years so why is there still resistance to putting automation in place?

**Comment by untitaker** [comment]
@cclauss I would not like flake8 in an entirely different configuration than what is already running for 2.7. I see you're updating sentry-flake8 too so I think it will be fine.

**Comment by cclauss** [comment]
I do not know how to fix `sentry-flake8`.

File: tests/sentry/integrations/github_enterprise/testutils.py
Comment (joshuarli) [inline tests/sentry/integrations/github_enterprise/testutils.py]: @cclauss why this change? Actually this prompted me to take a brief look, and I've noticed a mixture of byte string and raw string literals in `tests/sentry/integrations`. @MeredithAnya do you know anything about why some are one kind and some are the other?

File: .travis.yml
Comment (joshuarli) [inline .travis.yml]: @cclauss Why are you explicitly selecting these four? We also don't really need `--count --show-source --statistics`, reason being it makes the output pretty verbose.

**Comment by joshuarli** [comment]
Hmm, unsure why F841 and E211 appear on py3.7 sentry-flake8 only.

File: tests/sentry/integrations/github_enterprise/testutils.py
Comment (MeredithAnya) [reply tests/sentry/integrations/github_enterprise/testutils.py]: interesting https://github.com/getsentry/sentry/pull/14924/files originally was r

**Comment by mattrobenolt** [comment]
Why did security get roped into this?

**Comment by joshuarli** [comment]
Dunno.

File: tests/acceptance/conftest.py
Comment (mattrobenolt) [inline tests/acceptance/conftest.py]: If this changed to `str.format()` we need to change it to an unicode literal. Or just leave it how it was.

Granted, the change here would be fine, but I tried to make a pass of removing all `str.format()` calls everywhere since it causes problems in python2.

**Comment by mattrobenolt** [comment]
Oh, I bet because of this rule: https://github.com/getsentry/sentry/blob/master/.github/CODEOWNERS#L16

File: .travis.yml
Comment (cclauss) [reply .travis.yml]: F9 enables all F9xx tests.

https://flake8.pycqa.org/en/latest/user/error-codes.html

On the flake8 test selection, this PR does _not_ focus on "_style violations_" (the majority of flake8 error codes that [__psf/black__](https://github.com/psf/black) can autocorrect).  Instead these tests are focus on runtime safety and correctness:
* E9 tests are about Python syntax errors usually raised because flake8 can not build an Abstract Syntax Tree (AST).  Often these issues are a sign of unused code or code that has not been ported to Python 3.  These would be compile-time errors in a compiled language but in a dynamic language like Python they result in the script halting/crashing on the user.
* F63 tests are usually about the confusion between identity and equality in Python.  Use ==/!= to compare str, bytes, and int literals is the classic case.  These are areas where __a == b__ is True but __a is b__ is False (or vice versa).  Python >= 3.8 will raise SyntaxWarnings on these instances.
* F7 tests logic errors and syntax errors in type hints
* F82 tests are almost always _undefined names_ which are usually a sign of a typo, missing imports, or code that has not been ported to Python 3.  These also would be compile-time errors in a compiled language but in Python a __NameError__ is raised which will halt/crash the script on the user.

I would usually reject the PR if one of these errors was present so the size of helpful output is not a concern.  If these test pass then the output is one line which can be eliminated by removing --count.

File: tests/acceptance/conftest.py
Comment (cclauss) [reply tests/acceptance/conftest.py]: > str.format() calls everywhere since it causes problems in python2.

Can you please provide an example?

**Comment by cclauss** [comment]
https://travis-ci.org/getsentry/sentry/jobs/601023986#L366 is an example where --show-source would enhance developer productivity.

File: tests/acceptance/conftest.py
Comment (dcramer) [reply tests/acceptance/conftest.py]: Matt is talking about how unicode is implicitly coerced with str.format in Python 2.

'foo'.format(u'insert non ascii')

vs

u'foo'.format('insert non ascii')

Comment (mattrobenolt) [reply tests/acceptance/conftest.py]:
```python
"{}".format(u"ã")
```

We’ve historically had dozens of errors related to str.format vs unicode.format. I understand this isn’t a problem in python3 since they gratefully and thankfully removed the ability to format bytestrings. And keep it only to unicodes. But until then, would prefer we be explicit on unicode strings when we use .format() to avoid issues.

**Comment by joshuarli** [comment]
> https://travis-ci.org/getsentry/sentry/jobs/601023986#L366 is an example where --show-source would enhance developer productivity.

Wdym? `./tests/acceptance/conftest.py:31:17` is enough useful context.

File: tests/acceptance/conftest.py
Comment (joshuarli) [inline tests/acceptance/conftest.py]: It was actually `black` that autoformatted this, but I guess it isn't enough to detect that that `__future__.print_function` was imported.

Also lol, TIL this works on 2.x and 3.x:

```
>>> print  ('asdf')
asdf
>>> print ('asdf')
asdf
>>> print('asdf')
asdf
```

File: tests/acceptance/conftest.py
Comment (joshuarli) [reply tests/acceptance/conftest.py]: e597c9c

File: .travis.yml
Comment (joshuarli) [reply .travis.yml]: Thanks for elaborating! I guess what I'm really asking here is are `E9,F63,F7,F82` in the default ignore list (and that'd be why you're selecting them explicitly here)? I couldn't find the default ignore list after some casual googling.

**Comment by joshuarli** [comment]
@cclauss This LGTM to merge, I was just wondering if you had any to say regarding https://github.com/getsentry/sentry/pull/14906#discussion_r337694274 since I basically just took over your branch and wrote on top of some of your changes.

File: .travis.yml
Comment (cclauss) [reply .travis.yml]: Thiese are in the select list (not ignore list) because they find issues of runtime safety.  You can find them in [Node.js](https://github.com/nodejs/node-gyp/blob/master/.travis.yml#L91), [GitHub Actions CI](https://github.com/actions/starter-workflows/blob/master/ci/python-package.yml#L28), [Internet Archive](https://github.com/internetarchive/openlibrary/blob/master/Makefile#L96) so this was not a project-specific choice.  I hope they help.

**Comment by cclauss** [comment]
I added my comments above.  I think we are ready to go.

**Comment by joshuarli** [comment]
> I added my comments above. I think we are ready to go.

https://github.com/getsentry/sentry/pull/14906#discussion_r339728663

**Comment by joshuarli** [comment]
@lynnagara This doesn't look good, but it's also strange that flake8 under python2.7 doesn't catch this. F821 isn't ignored per our `setup.cfg` either.

```
./src/sentry/deletions/defaults/group.py:54:49: F821 undefined name 'event'
./src/sentry/deletions/defaults/group.py:55:44: F821 undefined name 'event'
```

Will be rebasing once your fix lands in master for this.

**Comment by untitaker** [comment]
>but it's also strange that flake8 under python2.7 doesn't catch this

`event` is leaked from the list comprehension in Python 2, while in Python 3 it isn't.

**Comment by untitaker** [comment]
https://github.com/getsentry/sentry/pull/15294

**Comment by joshuarli** [comment]
> > but it's also strange that flake8 under python2.7 doesn't catch this
> 
> `event` is leaked from the list comprehension in Python 2, while in Python 3 it isn't.

Ah yeah, I just remembered I discovered this at some point in the past, but didn't know it was python 2 specific.

File: .travis.yml
Comment (cclauss) [reply .travis.yml]: Correct.  If you have a pre-commit hooks or editors that runs black/yapf/autopep8 then just run `flake8 .` but if your contributors don’t have some automation then they will quickly get tired of their PRs being flagged for whitespace issues.  The Bxxx errors must come from https://github.com/PyCQA/bandit

Comment (untitaker) [reply .travis.yml]: B-something is from sentry-flake8.

Comment (joshuarli) [reply .travis.yml]: > Correct. If you have a pre-commit hooks or editors that runs black/yapf/autopep8 then just run `flake8 .` but if your contributors don’t have some automation then they will quickly get tired of their PRs being flagged for whitespace issues. The Bxxx errors must come from https://github.com/PyCQA/bandit

Very good point, although I think it would be better to emphasize in both dev docs and possibly in a pull request template to run the linter. We also do have pre-commit in place, which is installed via `make bootstrap`, but the linter hasn't been moved there yet for... reasons. I'll keep this in mind.

**Comment by cclauss** [comment]
The current undefined name errors are interesting ones.  In legacy Python `event` created in the list comprehension on line 50 would still be in scope on lines 54 and line 55.  However in Python 3, what happens in comprehension, stays in comprehension so `event` is outa scope.

**Review by wedamija — APPROVED** [review]

**Review by untitaker — APPROVED** [review]

**Review by dcramer — APPROVED** [review]
im surprised some of these werent caught by our existing lint rules

**Pull request merged**
